src/index.jl

```diff
@@ -439,7 +436,7 @@ Iterating over Index `I` gives the IndexVals `I(1)` through `I(dim(I))`.
 """
 function Base.iterate(i::Index, state::Int=1)
```

Comment:
I was actually meaning to get rid of this definition (#605), since I think by reading code like:
```julia
i = Index(5)
for x in i
  @show x
end
```
it would not be clear to me what `x` would be (i.e. you might guess it could be each pair `i => n` or it could be iterating over the values `n` themselves). I think it would be better to deprecate iteration of `Index` in favor of devoted iterator objects `eachval(::Index)` and `eachindval(::Index)`, for example:
```julia
eachval(i::Index) = 1:dim(i)
eachindval(i::Index) = (i => n for n in eachval(i))
```
Would you mind adding those? I think this isn't a straightforward case of deprecating with the `@deprecate` macro, so we could just keep the `Base.iterate` definition for now and use `Base.depwarn` inside it directly, for example just adding the line:
```julia
Base.depwarn("iteration of `Index` is deprecated, use `eachindval` or `eachval` instead.", :iterate)
```
We could just add a `println` but this ensures it gets turned on and off the same way as other deprecation warnings from `@deprecate`.

A bonus would be adding the same iterator definitions but for collections of indices:
```julia
eachval(is::Tuple{Vararg{Index}}) = CartesianIndices(dims(is))
eachindval(is::Tuple{Vararg{Index}}) = (is .=> Tuple(ns) for ns in eachval(is))
```
to allow for something like:
```julia
A = randomITensor(i, j, k)
B = randomITensor(i, x, k, j)
for ivs in eachindval((i, j, k))
  B[x => 1, ivs...] = A[ivs...]
end
```
similar to what was discussed here: #281. I believe there is a similar iterator defined in the C++ version.

Reply:
Yes good opportunity to add these, and they will be good to have. Just added them.